examples/huggingface/main.py

```diff
@@ -0,0 +1,320 @@
+        "-b",
+        "--per_device_train_batch_size",
+        help="Per device batch size",
+        type=int,
+        required=False,
+    )
+    parser.add_argument(
+        "-opt",
+        "--optim",
+        help="Optimizer (one of: adamw_hf, adamw_torch, adamw_apex_fused, or adafactor",
+        type=str,
+        required=False,
+    )
+    parser.add_argument(
+        "-wd",
+        "--weight_decay",
+        help="Weight decay for AdamW optimizer",
+        type=float,
+        required=False,
+    )
+    parser.add_argument(
+        "-b1",
+        "--adam_beta1",
+        help="beta1 hyperparameter for AdamW optimizer",
+        type=float,
+        required=False,
+    )
+    parser.add_argument(
+        "-b2",
+        "--adam_beta2",
+        help="beta2 hyperparameter for AdamW optimizer",
+        type=float,
+        required=False,
+    )
+    parser.add_argument(
+        "-eps",
+        "--adam_epsilon",
+        help="epsilon hyperparameter for AdamW optimizer",
+        type=float,
+        required=False,
+    )
+    parser.add_argument(
+        "-log", "--logfile", help="Log file name and path", type=str, required=False
+    )
+    args = parser.parse_args()
+    return vars(args)
+
+
+def set_training_args(training_args, args):
+    for argname, argvalue in args.items():
+        if argvalue is not None:
+            setattr(training_args, argname, argvalue)
+    return training_args
+
+
+class GPUMemoryCallback(TrainerCallback):
+    def on_epoch_end(self, args, state, control, **kwargs):
+        print(
+            "GPU mem: Tot - ",
+            torch.cuda.get_device_properties(0).total_memory,
+            "res - ",
+            torch.cuda.memory_reserved(0),
+            "used - ",
+            torch.cuda.memory_allocated(0),
+        )
+
+
+def get_free_gpu():
+    for i in range(torch.cuda.device_count()):
+        gpu_procs_str = torch.cuda.list_gpu_processes(i)
+        if "no processes are running" in gpu_procs_str:
+            return i
+    return None
+
+
+@hydra.main(config_path=".", config_name="config")
+def main(cfg: DictConfig) -> float:
+    print("args", cfg)
+
+    # Get command line arguments and apply hyperparaters to training arguments
+    args = cfg.args
+
+    # Logger setup
+    logfile = args["logfile"] or "translation_hf.log"
+
+    logging.basicConfig(filename=logfile, level=logging.INFO)
+    logger = logging.getLogger()
+
+    # Get a GPU if available
+    if torch.cuda.is_available():
+        device = f"cuda:{get_free_gpu()}"
+    else:
+        device = "cpu"
+
+    logger.info("Compute device: %s", device)
+
+    batch_size = 16
+
+    # Get hyperparameters
+    training_args = Seq2SeqTrainingArguments(
+        output_dir=str(os.getcwd()) + "/test_trainer",
+        save_total_limit=2,
+        per_device_train_batch_size=batch_size,
+        per_device_eval_batch_size=batch_size,
+        evaluation_strategy="epoch",
+        predict_with_generate=True,
```

Comment:
```suggestion
        predict_with_generate=True,
        **{key: value for key, value in args.items() if value is not None}
```
I think the following may be simpler than setting the args in training_args afterwards in a separate function.

Reply:
Some arguments that we have cant be created in a Seq2SeqTrainingArguments object. Throws the error : 
`training_args = Seq2SeqTrainingArguments(
TypeError: __init__() got an unexpected keyword argument 'size_train_dataset'`.
I think that when we use `setattr()` to add them after creating the object it sorts the correct ones and the ones we have for the general purpose of the code (A.K.A. changing the size for the datasets in the code).

What should I do here ?

Reply:
Oh so some are attributes but not arguments to the class constructor? That's a bit odd. Do they set attributes after creating the object in the tutoriel you are using?

Reply:
Yes, they first create the object than use `training_args = set_training_args(training_args, args)`

Reply:
I see. I cannot find `size_train_dataset` anywhere in TrainingArguments [here](https://github.com/huggingface/transformers/blob/main/src/transformers/training_args_seq2seq.py) and [here](https://github.com/huggingface/transformers/blob/main/src/transformers/training_args.py). I think `set_training_args` may be adding many unnecessary attributes to `training_args`. But it would be way to cumbersome to define them one by one as `TrainingArguments` supports many many many arguments (like 30+). So, let's keep it like that then.